### ISSUE TITLE
Some strange NaNs popping up during training

### ISSUE THREAD

**Issue by idanbeck** [huggingface/candle#1307]
Please dismiss this issue if it's just user error - I have not checked this against running the same data against a similar model in PyTorch but I've been doing some validation training on some data on a "toy" network and have been seeing NaNs cropping up after a certain amount of time.

The code below shows the forward pass and loss function, using the tanh on the last layer before the sigmoid seems to be resolving the NaNs in training (either SGD or Adam) so my guess is that maybe the activations are just getting really really large - so this well could be the issue although I haven't taken a minute to inspect the actual weights to confirm.

I'm including the loss, although I don't think this is the issue - I switched to an L2 norm loss, and am using max entropy as well since this toy model is meant to confirm this.

Figured I'd ask, and if there is a better place/forum for this kind of thing please let me know as this is less of a confirmed issue and more a request for a bit of clarification if I'm not using the library correctly (as mentioned, I haven't tried to repro on PyTorch or anything). Meanwhile, using the tanh instead of ReLU seems to resolve the NaNs in training.

`fn forward(&self, xs: &Tensor, train: bool) -> candle_core::Result {
let (b_sz, _img_dim) = xs.dims2()?;
let (w, h, c) = self.in_image_shape_whc;

    let xs = xs
        .reshape((b_sz, w, h, c))?
        .permute((0, 3, 1, 2))?
        .apply(&self.conv1)?
        .apply(&self.conv2)?
        .max_pool2d(2)?
        .apply(&self.conv3)?
        .apply(&self.conv4)?
        .max_pool2d(2)?
        .apply(&self.conv5)?
        .apply(&self.conv6)?
        .max_pool2d(2)?
        .apply(&self.conv7)?
        .max_pool2d(2)?
        .flatten_from(1)?
        .apply(&self.fc1)?;

    let xs = &xs.tanh()?;

    let xs = self.dropout.forward(&xs, train)?.apply(&self.fc2)?;
    //let xs = xs.apply(&self.fc2)?;
    let xs = sigmoid(&xs)?;

    return Ok(xs);
}`

`let lambda_maxent = (0.002 / batch_n as f64);
let log_states = output_states_eps.log()?;
let ent_states = (lambda_maxent * (-1.0 * output_states_eps.broadcast_mul(&log_states)?)?.sum_all()?)?;

let lambda_l2 = 1.0;
//let mse_loss = (lambda_mse * loss::mse(&output_states, &target_states)?)?;
let L2_loss = (lambda_l2 * (&output_states.powf(2.0)? - &target_states.powf(2.0)?)?
.abs()?
.mean_all()?
)?;
let loss = (L2_loss - ent_states)?;

    return Ok(loss);`

Thanks, and really excited about this library!

PS - SO sorry!  I had accidentally posted this issue originally in tch-rs, apologies for the confusion I might have caused

**Comment by idanbeck** [comment]
Resolved my own issues here - adding batch norm largely resolved this as I think the activations were exploding

Sorry for the distraction!
